File: NodeJS/errors/retry_failed_requests.js
const { chromium } = require('playwright');

const MAX_RETRIES = 3;

async function getPageHTML(url) {
    const browser = await chromium.launch();
    const page = await browser.newPage();

    for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
        try {
            await page.goto(url, { timeout: 10000 });
            const html = await page.content();
            await browser.close();
            return html;
        } catch (e) {
            console.log(`Attempt ${attempt} failed: ${e}`);
            await new Promise(r => setTimeout(r, 2000));
        }
    }

    await browser.close();
    throw new Error('All retries failed.');
}

getPageHTML("https://example.com").then(console.log);

File: Python/errors/retry_failed_requests.py
from playwright.sync_api import sync_playwright
import time

MAX_RETRIES = 3

def get_page_html(url):
    with sync_playwright() as p:
        browser = p.chromium.launch()
        page = browser.new_page()

        for attempt in range(1, MAX_RETRIES + 1):
            try:
                page.goto(url, timeout=10000)
                html = page.content()
                browser.close()
                return html
            except Exception as e:
                print(f"Attempt {attempt} failed: {e}")
                time.sleep(2)
        browser.close()
        raise Exception("All retries failed.")

html = get_page_html("https://example.com")
print(html)
